### ISSUE TITLE
UBSan complains of integer overflow at large sizes

### ISSUE THREAD

**Issue by bungeman** [harfbuzz/harfbuzz#3932]
Originally reported as [Chromium 48531](https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=48531). Essentially, Chromium's fuzzers for HarfBuzz use part of the data to select a font size (it appears that HarfBuzz's fuzzers do not, opting instead to use some fixed size). The fuzzer discovered that it could set the size to something very large like 0x7fffff80 and then UBSan complains about lots of signed integer overflow in HarfBuzz. Basically any place a signed integer is used in any math involving the size.

This does not appear to be a very urgent issue to fix. Even with this limitation it appears there shouldn't be any overflow in any sane font until at least one million points (~350 kilometers). When the overflow does happen it appears to only cause points to be in valid positions, though perhaps far from where one would expect.

However, this issue may be at least somewhat interesting to consider, as users cannot reasonably be expected to enforce any size constraints, but the library implementation may. For instance, it could simply be an error to set the size too large, which would be fine. More complex solutions like using floats, saturating arithmetic, or checking every bit of math are also possible but perhaps a bit much.

**Comment by behdad** [comment]
Yeah this is a known issue. We don't want to use saturating math everywhere. Using floats is my preference but breaks our ABI. Just limiting the font scale at entry point is my new preference that I had not thought about before.

**Comment by behdad** [comment]
> Just limiting the font scale

Humm. That's not possible because there's no number that will ensure there's no overflow. For example a font can have stacking diacritics...

**Comment by matthiasclasen** [comment]
Having some limit for font scale would still be sufficient to keep the fuzzers from going there...

**Comment by behdad** [comment]
A font scale of 0xFFFF is very reasonable thing to use so I don't want to put the limit lower than say 0x00FFFFFF. If there the fuzzer can still find very tall glyphs and stack them and hit the limit.

**Comment by behdad** [comment]
I'm fine limiting scale to upem*0xFFFF.

**Comment by behdad** [comment]
And stack only 16 marks max.

**Comment by behdad** [comment]
I don't see anything actionable here for now.
